[Bug 1963773][Harmony] login-names separate from email addresses

topunix opened the pull request:
#### Details

This PR is a draft of current work

#### Additional info
* [1963773](https://bugzilla.mozilla.org/show_bug.cgi?id=1963773)

topunix committed:
Adding test change

topunix committed:
Adding new auth, and new_login to Account page

topunix committed:
Added new user errors and save new_login to userprefs.cgi

topunix committed:
Added error check, login_change_during_email_change

topunix committed:
updated small login field

topunix committed:
changed loginname to login in placeholder

topunix committed:
Updated login input

topunix committed:
Added login_too_long error

topunix committed:
Improved error checking in check_login_name_for_creation

topunix committed:
removed email change logic

topunix committed:
Change current_tab to email

topunix committed:
Change current_tab to email

topunix committed:
Adding hook file for email tab

topunix committed:
Adding hook file for email tab

topunix committed:
Removing account-field

topunix committed:
Removing account-field

topunix committed:
Rename email-field file to account-field, reversing change

topunix committed:
Removing email-field.html.tmpl, reversing change

topunix committed:
Rename email-field file to account-field, reversing change

topunix committed:
Removing email-field.html.tmpl, reversing change

topunix committed:
Reversed tab check to account

topunix committed:
Reversed tab check to account

topunix committed:
Returning email address field

topunix committed:
Revert changes to userprefs.cgi from 4205bd7d00c61ef7a9bbf7aec0a8a198af98a8bb

topunix committed:
Adding secondary email placeholder

topunix committed:
Adding profiles_emails table

topunix committed:
Change column name

topunix committed:
Added new column

topunix committed:
Adding more test fields for emails

topunix committed:
Added login name to answers

topunix committed:
Adding Bugzilla::User::Email

topunix committed:
Adding creation of email account to Bugzilla::User::create()

topunix committed:
Removed unneeded columns

topunix committed:
Added validators and columns

topunix committed:
Adding email widget

topunix committed:
Enhanced email widget

topunix committed:
Added mutators, accessors, update, and check_user_id validator

topunix committed:
Added check email validator

topunix committed:
Added remove_from_db

topunix committed:
Updated button styling

topunix committed:
Adding display_order column

topunix committed:
Added get_user_emails subroutine

topunix committed:
added display_order to update_columns

topunix committed:
Change to remove_from_db

topunix committed:
Added reIndexEmailFields() and inline comments

topunix committed:
Updated input attributes for primary email

topunix committed:
Added display_order constants

topunix committed:
Fixed check_email validator

topunix committed:
Added find_user_by_email()

topunix committed:
Renamed subroutine

topunix committed:
Renamed subroutine for parity

topunix committed:
Converted to methods

topunix committed:
Added new(), validator, and name_field

topunix committed:
Added install_admin_get_login_name

topunix committed:
Added email_exists

topunix committed:
Improved validator

topunix committed:
Added Bugzilla::Error

topunix committed:
Added email to admin install process

topunix committed:
Fixed bug

topunix committed:
Fixed comment

topunix committed:
removed comment

topunix committed:
Added get_primary_email_of_user()

topunix committed:
Updated email accessor to use Bugzilla::User::Email

topunix committed:
Replaced email with login_name

topunix committed:
Revert account.html.tmpl to version from 07ece2b

topunix committed:
Added login_name and new_email fields

topunix committed:
Added new_login_name and new_email

topunix committed:
Added email_changed

topunix committed:
Updated email_exists

topunix committed:
Updated account_exists

topunix committed:
Replaced key name in user error

topunix committed:
Updated changeEmail() and confirm_create_account()

topunix committed:
Added email creation

topunix committed:
Replaced username with login, added email to create_or_update_user()

topunix committed:
Added set_email()

topunix committed:
Added email to user create

topunix committed:
Added login_name fallback to set_email

topunix committed:
Expand token data unpacking

topunix committed:
Added email param to issue_new_user_account_token()

topunix committed:
Added email param

topunix committed:
Added email var

topunix committed:
Added email string

topunix committed:
Updated template to use email in form

topunix committed:
Added email param to check_and_send_account_creation_confirmation()

topunix committed:
Added email to user error

topunix committed:
Replace emailaddress with login for User

topunix committed:
Added userid check for emailaddress and login vars

topunix committed:
Fallback to user login if primary email is undefined

topunix committed:
Update password reset error message to require email address

topunix committed:
Added ACCOUNT_CHANGE_INTERVAL constant

topunix committed:
Fix: Update password change error message and condition

topunix committed:
Fixed error name

topunix committed:
Refactor: Standardize on email variable for password requests and update user lookup

topunix committed:
Removed SEND_NOW

topunix committed:
Refactor: Implement user lookup when 'name' parameter is present

topunix committed:
Added invalid_email error

topunix committed:
Refactor: Simplify user instantiation and rely on constructor for validation

topunix committed:
Added email syntax check to password reset

topunix committed:
Removed email syntax check in constructor to allow non-email logins

topunix committed:
Added email address to search options

topunix committed:
Added email column

topunix committed:
feat: Add email to user profile query

topunix committed:
Added email to user account

topunix committed:
Prevented error with unrecognized column (email)

topunix committed:
Replace user with otheruser

topunix committed:
Docs: Update login name and email field help text

topunix committed:
Added _copy_valid_emails_to_profiles_emails()

topunix committed:
Fix: Handle undefined email address in validation

topunix committed:
Fix: Improve email population and validation in DB install

topunix committed:
Refactor: Handle optional 'email' column in profiles table during DB install

topunix committed:
Fix: Log 'unknown' if remote IP is not available during user creation audit

eyalroz commented:
I realize this question may be out-of-scope for this particular PR, but - where can we learn about the current state of the work on Bugzilla 6 / Harmony? What's been done over the past, say, year, and what are the prospects for completion/release?

justdave commented:
> I realize this question may be out-of-scope for this particular PR, but - where can we learn about the current state of the work on Bugzilla 6 / Harmony? What's been done over the past, say, year, and what are the prospects for completion/release?

This PR is one of the last BIG things left towards getting the release out. There are a number of smaller things still to go, but those shouldn't take as long. My goal is to get a blog post up within the next week with a current status. The main delay is lack of developer time. We're raising money to be able to pay a developer but aren't anywhere near the amount needed to do so yet, and  the few volunteers we have right now don't have a lot of time available.

topunix committed:
Removed tabs to resolve whitespace error from test file

topunix committed:
Added admin_login_name

eyalroz commented:
@justdave : That sounds promising, thank you :-)

So, would you say a release of 6.0 in 2025 is a realistic expectation?

Also, has a draft of the 6.0 release notes been completed? That's one of the release blockers, and you said everything except for the migration part was done.

justdave commented:
> So, would you say a release of 6.0 in 2025 is a realistic expectation?

Yes. 

> Also, has a draft of the 6.0 release notes been completed? That's one of the release blockers, and you said everything except for the migration part was done.

In my brain I always think of that as part of the release process so that slipped my mind talking about it. I made a lot of progress on that back in the November to January time frame, but it's not done yet.  The in-progress notes are in a Google spreadsheet, I don't remember if I've posted the link to it anywhere. I  just added a link to it on the [release notes bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1786958). I worked on it on-stream on the Twitch channel when I did it.

mrenvoize commented:
Those draft notes make for interesting reading.. lots removed.

justdave committed:
Merge branch 'main' into main

eyalroz commented:
> > So, would you say a release of 6.0 in 2025 is a realistic expectation?
> 
> Yes.

So, I noticed the merge... would you say you are on track for a release in 2025? Many people are waiting holding their fingers crossed :-)

justdave commented:
We should have 5.9.2 in the next week or two, with this in it if I don't find any problems with it. As this was the main remaining blocker to releasing Bugzilla 6, a release candidate should follow not long after.

justdave committed:
Merge branch 'main' into main

eyalroz commented:
> We should have 5.9.2 in the next week or two, with this in it if I don't find any problems with it. As this was the main remaining blocker to releasing Bugzilla 6, a release candidate should follow not long after.

So, 2025 is ending, and I have not noticed a 5.9.2 tag. But - I have noticed the merge of this PR. Could you update your time estimates from September?

justdave commented:
> > We should have 5.9.2 in the next week or two, with this in it if I don't find any problems with it. As this was the main remaining blocker to releasing Bugzilla 6, a release candidate should follow not long after.
> 
> So, 2025 is ending, and I have not noticed a 5.9.2 tag. But - I have noticed the merge of this PR. Could you update your time estimates from September?

It is impossible to make time estimates with only volunteers available as everything depends on when people can make time for Bugzilla outside of their primary income work. This is why the Project is trying to raise money to hire someone. That said, there is now less work remaining to do than there was in the last statement above, but work still remains, and I'd still like to get 5.9.2 out in the next week or two. But that depends on having time available to do it.  Which hopefully will happen, but I'm not in a position to make any guarantees.

justdave commented on template/en/default/admin/users/userdata.html.tmpl:
```diff
@@ -113,7 +129,7 @@
       <input type="checkbox" name="reset_bounce" id="reset_bounce" value="1">
-      (<a href="/bounced_emails/[% otheruser.id FILTER uri %]" title="View bounce email history">
+      (<a href="/bounced_emails/[% user.id FILTER uri %]" title="View bounce email history">
         [% otheruser.bounce_count FILTER html %]</a>)
```
This suggested change removes two lines and only replaces one of them, and the second one is still necessary. Line 133 needs to stay.

topunix replied:
Fixed in ae26661d7. Applied by hand rather than via the suggestion block, so line 133 is still there. Current state is otheruser.id on the href and otheruser.bounce_count on the following line.

justdave commented on Bugzilla/User/Email.pm:
```diff
@@ -0,0 +1,166 @@
+  validate_email_syntax($email)
+    || ThrowUserError('illegal_email_address', {addr => $email});
+  
+  if ($invocant->get_user_by_email($email)) {
+    ThrowUserError("email_exists", {'email' => $email})
+  }
```
I think Mike mentioned in his followup comment but this is actually NOT a syntax error because it's the last statement inside the braces {} and Perl only needs a semicolon between statements, not after every one. But it won't break anything to have it, and it'll make potential future changes less likely to break things if someone eventually adds another line after it and forgets to add it then.

topunix replied:
Added in 7d7400049. Note this was not a syntax error, since it was the last statement in the block, so nothing was actually broken.

justdave commented on template/en/default/global/user-error.html.tmpl:
```diff
@@ -72,7 +72,7 @@
     [% title = "Account Already Exists" %]
     There is already an account with
     [% IF email %]
-      the login name [% email FILTER html %].
+      the login name [% login FILTER html %].
     [% ELSE %]
       that login name.
     [% END %]
```
Suggested fix here, since it didn't give one, the [% IF email %] should be [% IF login %] but call sites also need to be double-checked to make sure that's what they're passing.

topunix replied:
Fixed in 7d7400049. Checked the call sites as suggested: Bugzilla/User.pm:351 in check_login_name_for_creation is the only place that throws this error, and it passes {login => $name}, so it matches. The other hit for the string is token.cgi:299, which passes account_exists as a cancel action to Bugzilla::Token::Cancel rather than as a user error. That routes to cancel-token.txt.tmpl, which has its own branch using email, and that one is correct since the collision there is on the email address.

topunix committed:
Bug 1963773: Add method=post to account creation form

topunix committed:
Bug 1963773: Replace MySQL-only SHOW COLUMNS with bz_column_info

topunix committed:
Bug 1963773: Fix shadowed  in token.cgi reqpw

topunix commented on token.cgi:
```diff
@@ -101,15 +102,19 @@ if ($action eq 'reqpw') {
+
+  unless ($user_account) {
+    ThrowUserError("invalid_email", {email => $email});
+  }
```
Fixed in 6fa78fbed

topunix commented on template/en/default/account/create.html.tmpl:
```diff
@@ -1,67 +1,74 @@
 [% END %]
 
-[% PROCESS "account/auth/signup-form.html.tmpl" %]
+<form id="account_creation_form" action="createaccount.cgi" class="table_layout">
```
Fixed in f66c89d85

topunix commented on Bugzilla/Install/DB.pm:
```diff
@@ -4393,6 +4394,64 @@ sub _populate_attachment_storage_class {
+    # Check if 'email' column exists in 'profiles'
+    my $columns = $dbh->selectcol_arrayref(
+        "SHOW COLUMNS FROM profiles LIKE 'email'"
+    );
+    my $has_email_column = scalar(@$columns) > 0;
+
```
Fixed in 587df9e3b

justdave committed:
Merge branch 'main' into main

topunix committed:
Bug 1963773: fix addr/login/email reporting in three Copilot review items

- Auth/Verify.pm: report addr => $email instead of $login in auth_invalid_email error
- user-error.html.tmpl: account_exists now checks IF login instead of IF email
- Email.pm: add missing trailing semicolon after ThrowUserError in check_email_for_creation

topunix commented on Bugzilla/Install/DB.pm:
```diff
@@ -817,7 +817,8 @@ sub update_table_definitions {
   _populate_attachment_storage_class();
 
-
+  # Bug 1963773 - topunixguy@gmail.com
+  _copy_valid_emails_to_profiles_emails(); 
```
consistent with file convention

topunix committed:
Bug 1963773: fix invalid_email wording, token.cgi hashref bug, email validation guard

- user-error.html.tmpl: reword invalid_email as not registered, not invalid
- token.cgi: fix Bugzilla::User::Email->new() call to use hashref (was silently dropping old_email)
- Config/Common.pm: fix defined guard so unparseable input is rejected, not silently passed

topunix committed:
Bug 1963773: fix regex injection and admin/otheruser field mixups
- Token.pm: quotemeta($email) before building regexp to prevent metacharacter injection
- userdata.html.tmpl: bounce link now uses otheruser.id instead of user.id
- userdata.html.tmpl: disabledtext now defaults from otheruser instead of user

topunix committed:
Bug 1963773: Fix editusers.cgi user list JOIN

Change INNER JOIN to LEFT JOIN on profiles_emails so users without
an email row still appear. Add is_primary_email = 1 to the join
condition to prevent duplicate rows per user.

topunix committed:
Bug 1963773: Wire email param read/persist in editusers.cgi

The admin user form posts an email field but editusers.cgi never read
it, so the value was silently discarded on submit.

Pass email through on account creation, and on update persist the
primary address in profiles_emails: update the existing primary row if
one exists, otherwise create it. Validates only when the value actually
changed so unchanged resubmits do not trip email_exists.

Uses Bugzilla::User::Email directly rather than User->set_email, which
still blind-inserts and would violate the unique index on update.

topunix commented on editusers.cgi:
```diff
@@ -77,15 +77,28 @@ elsif ($action eq 'list') {
 
+  my $select_fields = 'profiles.userid, profiles.login_name, profiles.realname, profiles.is_enabled, '
+                  . $dbh->sql_date_format('profiles.last_seen_date', '%Y-%m-%d') . ' AS last_seen_date';
+
+  # Add email as a column from profiles_emails table
+  $select_fields .= ', profiles_emails.email AS email';
+
+  my $query = 'SELECT DISTINCT ' . $select_fields . ' FROM profiles';
+  
+  # Join the two tables by userid
+  $query .= ' INNER JOIN profiles_emails ON profiles.userid = profiles_emails.user_id';
+
```
Fixed in e745356f366e9faa67cc9664caabffc5616a0c7d

topunix committed:
Bug 1963773: Add email branch to account_updated message

The changed_fields loop in the account_updated message is a fixed
IF/ELSIF chain. Persisting the email address in editusers.cgi adds an
email key to the changes hash, which rendered as an empty list item
because no branch matched. Add a branch reporting the new address.

topunix committed:
Bug 1963773: Require Perl 5.14.0 in Bugzilla::User::Email

The new module predates the 5.14.0 sweep in bug 2052697 and still
declared 5.10.1, which fails t/002goodperl.t.

topunix commented on Bugzilla/Auth/Verify.pm:
```diff
@@ -55,51 +56,52 @@ sub create_or_update_user {
-    if ($username && !$username_user_id && !$extern_user_id) {
-      validate_email_syntax($username) || return {
+    if ($login && !$login_user_id && !$extern_user_id) {
+      validate_email_syntax($email) || return {
         failure => AUTH_ERROR,
         error   => 'auth_invalid_email',
-        details => {addr => $username}
+        details => {addr => $login}
       };
```
Fixed in 7d7400049.

topunix commented on Bugzilla/Config/Common.pm:
```diff
@@ -79,7 +79,7 @@ sub check_regexp {
 sub check_email {
   my ($value) = @_;
   my ($address) = Email::Address::XS->parse($value);
-  if (!$address->is_valid) {
+  if (defined $address && !$address->is_valid) {
     return "must be a valid email address.";
   }
```
Fixed in 3809beb78.

topunix committed:
Bug 1963773: Skip regexp validation on machine-built token pattern

The throttling pattern is built with quotemeta and cannot be invalid,
so bz_check_regexp only costs a round-trip query, and it validates the
SQL-quoted string rather than the pattern itself. Pass nocheck.

topunix commented on Bugzilla/Token.pm:
```diff
@@ -96,44 +96,44 @@ sub check_auth_delegation_token {
           WHERE tokentype = ?
-                AND ' . $dbh->sql_istrcmp('eventdata', '?') . '
+                AND ' . $dbh->sql_regexp('eventdata', $dbh->quote($regexp)) . '
                 AND issuedate > '
```
Fixed in ae26661d7 (quotemeta) and c20dea825 (nocheck on sql_regexp).

topunix committed:
Bug 1963773: Remove tab characters from Token.pm

Fixes t/005whitespace.t.

topunix commented on scripts/entrypoint.pl:
```diff
@@ -123,14 +123,14 @@ sub cmd_dev_httpd {
     run(
       'perl', 'scripts/generate_bmo_data.pl',
       '--param' => 'use_mailer_queue=0',
-      'admin@bmo.test'
+      'admin_bmo_test'
     );
```
Verified, this works. Bugzilla::User::new now resolves a {name => ...} lookup against profiles_emails first and falls back to login_name when no email row matches, so passing a login name here finds the admin. Confirmed against a running instance: generate_bmo_data.pl seeds correctly with admin_bmo_test.

topunix commented on template/en/default/admin/users/userdata.html.tmpl:
```diff
@@ -38,6 +38,22 @@
     [% END %]
   </td>
 </tr>
+<tr>
+<th> <label for="email"> Email address: </label> </th>
+<td>
+    [% IF editusers %]
+      <input size="64" maxlength="255" name="email"
+             id="email" value="[% otheruser.email FILTER html %]">
+      [% IF editform %]
+        [% IF !otheruser.in_group('bz_sudo_protect') %]
+          <br>
+        [% END %]
+      [% END %]
+    [% ELSE %]
+      [% otheruser.email FILTER html %]
+    [% END %]
+</td>
+</tr>
```
Fixed in e745356f3 (read and persist the email param on create and update) and 00555af42 (account_updated message needed an email branch, since the new changes key rendered as an empty list item).

topunix commented on template/en/default/admin/users/userdata.html.tmpl:
```diff
@@ -128,7 +144,7 @@
          maxrows        = 10
          defaultrows    = 10
          cols           = 60
-         defaultcontent = otheruser.disabledtext
+         defaultcontent = user.disabledtext
        %]<br>
```
Fixed in ae26661d7.

topunix commented on template/en/default/global/user-error.html.tmpl:
```diff
@@ -1109,6 +1118,12 @@
     [% title = "Invalid Flag ID" %]
     The flag id [% flag_id FILTER html %] is invalid.
 
+  [% ELSIF error == "invalid_email" %]
+    [% title = "Invalid Email Address" %]
+    The address <tt>[% email FILTER html %]</tt> is not a valid email address.
+    Either you misspelled it, or the person has not
+    registered for a [% terms.Bugzilla %] account.
+
```
Fixed in 3809beb78.

topunix commented on token.cgi:
```diff
@@ -289,21 +294,20 @@ sub changeEmail {
+  # Update the user's email address and delete the token
   # from the tokens table.
+
+  my $user_email = Bugzilla::User::Email->new('name' => $old_email);
```
Fixed in 3809beb78.

topunix committed:
Bug 1963773: Make User::set_email upsert instead of blind insert

set_email always created a new profiles_emails row marked primary, so
any update path violated the unique index on email or left two rows
flagged primary. Update the existing primary row when one is present,
create it otherwise, and return early when the address is unchanged.

topunix commented on Bugzilla/User.pm:
```diff
@@ -434,6 +451,19 @@ sub _generate_nickname {
   return $nick;
 }
 
+sub set_email {
+  my ($self, $email) = @_;
+  # Create a user email account
+  my $email_data = {
+    user_id          => $self->id,
+    email            => $email,
+    is_primary_email => 1
+  };
+
+  my $user_email = Bugzilla::User::Email->create($email_data);
+}
```
Fixed in 03008ad78. set_email now updates the existing primary row when one is present, creates it when there isn't, and returns early when the address is unchanged. Verified against a running instance: repeated set_email calls update the same profiles_emails row in place, one row per user, no constraint violation.

topunix committed:
Bug 1963773: Parse email:login eventdata in remaining token.cgi paths

issue_new_user_account_token now stores eventdata as "email:login", but
request_create_account and cancel_create_account still treated it as a
single value, so the confirmation page showed the raw combined string
as the email address.

Also stop substituting a CGI-supplied login when the token login is
already taken. The confirmation form has no login field, so the
substitution always yielded undef. Cancel the token and report the
conflict instead.

Show the login name on the confirmation page, which previously only
displayed the email address.

topunix commented on token.cgi:
```diff
@@ -419,8 +423,13 @@ sub confirm_create_account {
   Bugzilla->assert_password_is_secure($password1);
   Bugzilla->assert_passwords_match($password1, $password2);
 
+  # Be careful! Some logins may contain ":" in them.
+  my ($email, $login) = split(':', $data, 2);
+  $login = $cgi->param('login') if login_to_id($login);
+
```
Fixed in e5cc69c3a. Worth noting the substitution never actually worked: confirm-new.html.tmpl has no login field, so $cgi->param('login') was always undef and a taken login produced a user_login_required error rather than an arbitrary login. Now the conflict is rejected directly, cancelling the token and reporting account_exists, so the user gets a clear message and a dead token isn't left behind to fail the same way on retry.

topunix commented on Bugzilla/Token.pm:
```diff
@@ -96,44 +96,44 @@ sub check_auth_delegation_token {
+    ThrowUserError('too_soon_for_new_token', {'type' => 'account'}) if $pending_requests;
 
-  my ($token, $token_ts) = _create_token(undef, 'account', $login_name);
+    my ($token, $token_ts) = _create_token(undef, 'account', "$email:$login");
```
Fixed in e5cc69c3a. Both paths now split eventdata on the first colon, matching what Token::Cancel and confirm_create_account already did. request_create_account also drops the emailsuffix concatenation, since the address is stored whole now, and passes the login through so the confirmation page can show it. Checked the remaining GetTokenData callers: the others only use the userid, so these two were the last consumers reading eventdata as a single value.

justdave committed:
Merge branch 'main' into main

justdave commented on Bugzilla/Token.pm:
`IssueEmailChangeToken()` sets `$old_email` from `$user->login` and sends the tokens to that value, and the account prefs path calls it directly when saving a new email (`userprefs.cgi:166`). With login names now separated from email addresses, this will mis-send confirmation mail and generate tokens that cannot be confirmed for any account whose login is not also its email.

justdave commented on Bugzilla/User.pm:
```diff
@@ -154,6 +156,15 @@ sub new {
       $param = {condition => 'extern_id = ?', values => [$param->{extern_id}]};
       $_[0] = $param;
     }
+    elsif (exists $param->{name}) {
+      my $email = $param->{name};
+      
+      my $user_id = Bugzilla::User::Email->get_user_by_email($email);
+      if ($user_id) {
+        $param->{id} = $user_id;
+        delete $param->{name};
+      }
+    }
```
`Bugzilla::User->new({ name => ... })` now prefers an email-row match over the login-name match. Good idea in theory, but it changes the meaning of a very widely used constructor and makes lookups ambiguous whenever a login happens to equal some other account’s email. Existing login-based callers will silently get the wrong account in that case, so this needs either a separate email lookup path or preserved login-first semantics. Personally I think this'll be safer to be explicit about which one you're trying to look up from all callers. The idea that someone could possibly set their username to an email address that belongs to someone else bothers me, too. More on that in a general review comment.

justdave reviewed (changes requested):
In addition to the code-referenced review comments, something jumped out at me as I was looking this over:

A username can be set to an email-looking string that belongs to someone else’s account, because the login uniqueness check only looks at profiles.login_name via login_to_id() and does not compare against profiles_emails at all. That means the underlying check in User.pm will not reject someone else’s email unless it is also already a login name. The self-service UI in `userprefs.cgi` blocks `@` in the login field, so regular users can’t do it there, but admin/backend paths still can. Perhaps this part can be dealt with by also blocking `@` in login names on the back end (for any changes made, not existing ones grandfathered in) so that admins can't do it either. (or let an admin do it with a warning?)

Likewise, adding an email address that matches someone else’s username is allowed in the current logic, because `Bugzilla::User::Email->check_email_for_creation()` only checks email syntax plus email uniqueness in `profiles_emails`. It does not check whether that value is already used as a login name anywhere. This one is probably worth checking to make sure someone can't add an email address that's in someone else's grandfathered-in login name.

justdave commented:
> For the backend block, do you want a hard reject on any new or changed login containing @, or an admin override with a warning? Happy either way, just want to match what you'd prefer before I write it. Existing logins stay grandfathered in either case.

Let's hard block it on new or changed. Just make sure it gets documented.

topunix commented:
> Let's hard block it on new or changed. Just make sure it gets documented.

Done in db20fa2ca. Hard block on new or changed logins, existing ones grandfathered, rule documented in the admin guide. Also added the reverse check so a new email can't collide with a grandfathered login. Same commit adds a client side pattern on the login field and lines the two form labels up.

topunix commented:
The `@` block in db20fa2ca conflicts with the ComponentWatching extension, which is why CI is currently red.

 `_check_watch_user` (extensions/ComponentWatching/Extension.pm:250) requires the watch user to match `/@(?!invalid).+\.bugs$/i`, and `_generate_watch_user_name` builds `component@product.bugs`. Three paths then push that value into `login_name`: the hardcoded `testcomponent@testproduct.bugs` at line 172, `_create_watch_user` at line 237, and the rename path at line 205 which calls `set_login`. All three now throw.

That means component creation and component rename are broken on any install running the extension, so this is not just a test fixture issue. The five failing test files are all downstream of the line 172 create.

Options as I see them:

 1. Rework ComponentWatching so the generated address becomes the account email and the login is a sanitized non-`@` form. Largest of the three, and it touches a template with JavaScript that the comment at line 216 says must be kept in sync with the Perl.
 2. Exempt auto-generated system accounts from the rule.
 3. Narrow the block to human-facing paths rather than the validator.

Which direction do you want? I'd rather get your call on the shape before going further, since option 1 is substantially larger than the other two.
